### PR TITLE
sshguard: T4408: Add service ssh dynamic-protection

### PR DESCRIPTION
**Pull request by sever-sever**
## Change Summary

Sshguard protects hosts from brute-force attacks
Can inspect logs and block "bad" addresses by threshold
Auto-generate rules for nftables
When service stopped all generated rules are deleted
nft "type filter hook input priority filter - 10"

Requires pkg `sshguard` from vyos-build

## Related PR
https://github.com/vyos/vyos-build/pull/233

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

* https://phabricator.vyos.net/T4408

## Component(s) name

ssh, sshguard
## How to test
Minimal configuration:
```
set service ssh dynamic-protection
```
Possible options:
```
set service ssh dynamic-protection block-time 120
set service ssh dynamic-protection detect-time 1800
set service ssh dynamic-protection threshold 30
set service ssh dynamic-protection whitelist-address 192.0.2.1
```

Check ruleset:
```
table ip sshguard {
	set attackers {
		type ipv4_addr
		flags interval
	}

	chain blacklist {
		type filter hook input priority filter - 10; policy accept;
		ip saddr @attackers drop
	}
}
table ip6 sshguard {
	set attackers {
		type ipv6_addr
		flags interval
	}

	chain blacklist {
		type filter hook input priority filter - 10; policy accept;
		ip6 saddr @attackers drop
	}
}
```
Check sshguard.config:
```
vyos@tstrtr2# cat /etc/sshguard/sshguard.conf | egrep -v "#"

BACKEND="/usr/libexec/sshguard/sshg-fw-nft-sets"

LOGREADER="LANG=C journalctl -afb -p info -n1 -t sshd -o cat"

THRESHOLD=30

BLOCK_TIME=120

DETECTION_TIME=1800

WHITELIST_FILE=/etc/sshguard/whitelist
```

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: interface-definitions/ssh.xml.in [hunk 1]
```diff
@@ -61,6 +61,80 @@
...
+                  <valueHelp>
+                    <format>ipv6</format>
+                    <description>IPv6 address to match against</description>
+                  </valueHelp>
+                  <valueHelp>
+                    <format>ipv6net</format>
+                    <description>IPv6 address and prefix length</description>
+                  </valueHelp>
+                  <constraint>
+                    <validator name="ipv4-address"/>
```
Comment (c-po) [inline interface-definitions/ssh.xml.in]: Could be shortened to `ip-address` and `ip-prefix`.

File: smoketest/scripts/cli/test_service_ssh.py [hunk 2]
```diff
@@ -213,5 +213,50 @@ def ssh_send_cmd(command, username, password, host='localhost'):
...
+        # Check configured port
+        tmp = get_config_value('Port')
+        self.assertIn(port, tmp)
+
+        # Check sshgurad service
+        self.assertTrue(process_named_running(SSHGUARD_PROCESS))
+
+        sshguard_lines = [
+            'THRESHOLD=10',
+            'BLOCK_TIME=123',
```
Comment (c-po) [inline smoketest/scripts/cli/test_service_ssh.py]: Please reference to the `block_time` variable above.

`f'BLOCK_TIME={block_time}',`

File: smoketest/scripts/cli/test_service_ssh.py [hunk 3]
```diff
@@ -213,5 +213,50 @@ def ssh_send_cmd(command, username, password, host='localhost'):
...
+        for line in sshguard_lines:
+            self.assertIn(line, tmp_sshguard_conf)
+
+        # Delete service ssh dynamic-protection
+        # but not service ssh itself
+        self.cli_delete(base_path + ['dynamic-protection'])
+        self.cli_commit()
+
+        self.assertFalse(process_named_running(SSHGUARD_PROCESS))
+        self.assertTrue(process_named_running(PROCESS_NAME))
```
Comment (c-po) [inline smoketest/scripts/cli/test_service_ssh.py]: This line is already tested in `tearDown()` which is invoked after this function exists.

**Review by c-po — CHANGES_REQUESTED** [review]

**Comment by c-po** [comment]
What happens on FW ruleset change? Will the sshguard lines persist or will the be wiped away?

**Comment by sever-sever** [comment]
> What happens on FW ruleset change? Will the sshguard lines persist or will the be wiped away?

After restart service all ip addresses are cleaned
```
vyos@tstrtr2:~$ cat test.txt | strip-private 
table ip sshguard {
	set attackers {
		type ipv4_addr
		flags interval
		elements = { xxx.xxx.241.146, xxx.xxx.45.227,
			     xxx.xxx.228.254, xxx.xxx.187.12,
			     xxx.xxx.65.33, xxx.xxx.65.126,
			     xxx.xxx.139.42, xxx.xxx.97.138,
			     xxx.xxx.172.60, xxx.xxx.172.61,
			     xxx.xxx.172.76, xxx.xxx.172.87,
			     xxx.xxx.172.89, xxx.xxx.172.91,
			     xxx.xxx.172.108, xxx.xxx.172.114,
			     xxx.xxx.172.124, xxx.xxx.172.160,
			     xxx.xxx.172.174, xxx.xxx.173.10,
			     xxx.xxx.173.16, xxx.xxx.173.35,
			     xxx.xxx.173.39, xxx.xxx.173.40,
			     xxx.xxx.173.43, xxx.xxx.173.44,
			     xxx.xxx.173.46, xxx.xxx.173.47,
			     xxx.xxx.173.48, xxx.xxx.173.49,
			     xxx.xxx.173.50, xxx.xxx.173.51,
			     xxx.xxx.173.52, xxx.xxx.173.53,
			     xxx.xxx.173.54, xxx.xxx.173.55,
			     xxx.xxx.173.56, xxx.xxx.173.61,
			     xxx.xxx.173.62, xxx.xxx.85.135,
			     xxx.xxx.85.237, xxx.xxx.64.194,
			     xxx.xxx.127.189, xxx.xxx.240.30,
			     xxx.xxx.17.54, xxx.xxx.163.64,
			     xxx.xxx.89.110, xxx.xxx.91.7,
			     xxx.xxx.126.221, xxx.xxx.96.99,
			     xxx.xxx.58.171, xxx.xxx.212.31,
			     xxx.xxx.216.128, xxx.xxx.71.137,
			     xxx.xxx.55.93, xxx.xxx.157.239,
			     xxx.xxx.167.196, xxx.xxx.50.147,
			     xxx.xxx.118.90, xxx.xxx.118.203,
			     xxx.xxx.10.157, xxx.xxx.10.174,
			     xxx.xxx.10.175, xxx.xxx.11.20,
			     xxx.xxx.11.29, xxx.xxx.25.38,
			     xxx.xxx.60.16, xxx.xxx.173.44,
			     xxx.xxx.154.134, xxx.xxx.154.203,
			     xxx.xxx.167.74, xxx.xxx.168.126,
			     xxx.xxx.151.130, xxx.xxx.117.222,
			     xxx.xxx.69.18, xxx.xxx.162.138,
			     xxx.xxx.0.221, xxx.xxx.202.149 }
	}

	chain blacklist {
		type filter hook input priority -10; policy accept;
		ip saddr @attackers drop
	}
}
```
Restart:
```
n1:~# systemctl restart sshguard
n1:~# 
n1:~# nft list table ip sshguard
table ip sshguard {
	set attackers {
		type ipv4_addr
		flags interval
	}

	chain blacklist {
		type filter hook input priority -10; policy accept;
		ip saddr @attackers drop
	}
}
```

File: interface-definitions/ssh.xml.in (hunk 1 above)
Comment (sever-sever) [reply interface-definitions/ssh.xml.in]: Done

File: smoketest/scripts/cli/test_service_ssh.py (hunk 2 above)
Comment (sever-sever) [reply smoketest/scripts/cli/test_service_ssh.py]: Done

File: smoketest/scripts/cli/test_service_ssh.py (hunk 3 above)
Comment (sever-sever) [reply smoketest/scripts/cli/test_service_ssh.py]: Done

**Comment by c-po** [comment]
> > What happens on FW ruleset change? Will the sshguard lines persist or will the be wiped away?
> 
> After restart service all ip addresses are cleaned
> 
> ```
> vyos@tstrtr2:~$ cat test.txt | strip-private 
> table ip sshguard {
> 	set attackers {
> 		type ipv4_addr
> 		flags interval
> 		elements = { xxx.xxx.241.146, xxx.xxx.45.227,
> 			     xxx.xxx.228.254, xxx.xxx.187.12,
> 			     xxx.xxx.65.33, xxx.xxx.65.126,
> 			     xxx.xxx.139.42, xxx.xxx.97.138,
> 			     xxx.xxx.172.60, xxx.xxx.172.61,
> 			     xxx.xxx.172.76, xxx.xxx.172.87,
> 			     xxx.xxx.172.89, xxx.xxx.172.91,
> 			     xxx.xxx.172.108, xxx.xxx.172.114,
> 			     xxx.xxx.172.124, xxx.xxx.172.160,
> 			     xxx.xxx.172.174, xxx.xxx.173.10,
> 			     xxx.xxx.173.16, xxx.xxx.173.35,
> 			     xxx.xxx.173.39, xxx.xxx.173.40,
> 			     xxx.xxx.173.43, xxx.xxx.173.44,
> 			     xxx.xxx.173.46, xxx.xxx.173.47,
> 			     xxx.xxx.173.48, xxx.xxx.173.49,
> 			     xxx.xxx.173.50, xxx.xxx.173.51,
> 			     xxx.xxx.173.52, xxx.xxx.173.53,
> 			     xxx.xxx.173.54, xxx.xxx.173.55,
> 			     xxx.xxx.173.56, xxx.xxx.173.61,
> 			     xxx.xxx.173.62, xxx.xxx.85.135,
> 			     xxx.xxx.85.237, xxx.xxx.64.194,
> 			     xxx.xxx.127.189, xxx.xxx.240.30,
> 			     xxx.xxx.17.54, xxx.xxx.163.64,
> 			     xxx.xxx.89.110, xxx.xxx.91.7,
> 			     xxx.xxx.126.221, xxx.xxx.96.99,
> 			     xxx.xxx.58.171, xxx.xxx.212.31,
> 			     xxx.xxx.216.128, xxx.xxx.71.137,
> 			     xxx.xxx.55.93, xxx.xxx.157.239,
> 			     xxx.xxx.167.196, xxx.xxx.50.147,
> 			     xxx.xxx.118.90, xxx.xxx.118.203,
> 			     xxx.xxx.10.157, xxx.xxx.10.174,
> 			     xxx.xxx.10.175, xxx.xxx.11.20,
> 			     xxx.xxx.11.29, xxx.xxx.25.38,
> 			     xxx.xxx.60.16, xxx.xxx.173.44,
> 			     xxx.xxx.154.134, xxx.xxx.154.203,
> 			     xxx.xxx.167.74, xxx.xxx.168.126,
> 			     xxx.xxx.151.130, xxx.xxx.117.222,
> 			     xxx.xxx.69.18, xxx.xxx.162.138,
> 			     xxx.xxx.0.221, xxx.xxx.202.149 }
> 	}
> 
> 	chain blacklist {
> 		type filter hook input priority -10; policy accept;
> 		ip saddr @attackers drop
> 	}
> }
> ```
> 
> Restart:
> 
> ```
> n1:~# systemctl restart sshguard
> n1:~# 
> n1:~# nft list table ip sshguard
> table ip sshguard {
> 	set attackers {
> 		type ipv4_addr
> 		flags interval
> 	}
> 
> 	chain blacklist {
> 		type filter hook input priority -10; policy accept;
> 		ip saddr @attackers drop
> 	}
> }
> ```

Actually I ment what happens when you delete the VyOS firewall - or update a rule. Will it influence sshguard?

**Comment by sever-sever** [comment]
> > > What happens on FW ruleset change? Will the sshguard lines persist or will the be wiped away?
> > 
> > 
> > After restart service all ip addresses are cleaned
> > ```
> > vyos@tstrtr2:~$ cat test.txt | strip-private 
> > table ip sshguard {
> > 	set attackers {
> > 		type ipv4_addr
> > 		flags interval
> > 		elements = { xxx.xxx.241.146, xxx.xxx.45.227,
> > 			     xxx.xxx.228.254, xxx.xxx.187.12,
> > 			     xxx.xxx.65.33, xxx.xxx.65.126,
> > 			     xxx.xxx.139.42, xxx.xxx.97.138,
> > 			     xxx.xxx.172.60, xxx.xxx.172.61,
> > 			     xxx.xxx.172.76, xxx.xxx.172.87,
> > 			     xxx.xxx.172.89, xxx.xxx.172.91,
> > 			     xxx.xxx.172.108, xxx.xxx.172.114,
> > 			     xxx.xxx.172.124, xxx.xxx.172.160,
> > 			     xxx.xxx.172.174, xxx.xxx.173.10,
> > 			     xxx.xxx.173.16, xxx.xxx.173.35,
> > 			     xxx.xxx.173.39, xxx.xxx.173.40,
> > 			     xxx.xxx.173.43, xxx.xxx.173.44,
> > 			     xxx.xxx.173.46, xxx.xxx.173.47,
> > 			     xxx.xxx.173.48, xxx.xxx.173.49,
> > 			     xxx.xxx.173.50, xxx.xxx.173.51,
> > 			     xxx.xxx.173.52, xxx.xxx.173.53,
> > 			     xxx.xxx.173.54, xxx.xxx.173.55,
> > 			     xxx.xxx.173.56, xxx.xxx.173.61,
> > 			     xxx.xxx.173.62, xxx.xxx.85.135,
> > 			     xxx.xxx.85.237, xxx.xxx.64.194,
> > 			     xxx.xxx.127.189, xxx.xxx.240.30,
> > 			     xxx.xxx.17.54, xxx.xxx.163.64,
> > 			     xxx.xxx.89.110, xxx.xxx.91.7,
> > 			     xxx.xxx.126.221, xxx.xxx.96.99,
> > 			     xxx.xxx.58.171, xxx.xxx.212.31,
> > 			     xxx.xxx.216.128, xxx.xxx.71.137,
> > 			     xxx.xxx.55.93, xxx.xxx.157.239,
> > 			     xxx.xxx.167.196, xxx.xxx.50.147,
> > 			     xxx.xxx.118.90, xxx.xxx.118.203,
> > 			     xxx.xxx.10.157, xxx.xxx.10.174,
> > 			     xxx.xxx.10.175, xxx.xxx.11.20,
> > 			     xxx.xxx.11.29, xxx.xxx.25.38,
> > 			     xxx.xxx.60.16, xxx.xxx.173.44,
> > 			     xxx.xxx.154.134, xxx.xxx.154.203,
> > 			     xxx.xxx.167.74, xxx.xxx.168.126,
> > 			     xxx.xxx.151.130, xxx.xxx.117.222,
> > 			     xxx.xxx.69.18, xxx.xxx.162.138,
> > 			     xxx.xxx.0.221, xxx.xxx.202.149 }
> > 	}
> > 
> > 	chain blacklist {
> > 		type filter hook input priority -10; policy accept;
> > 		ip saddr @attackers drop
> > 	}
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Restart:
> > ```
> > n1:~# systemctl restart sshguard
> > n1:~# 
> > n1:~# nft list table ip sshguard
> > table ip sshguard {
> > 	set attackers {
> > 		type ipv4_addr
> > 		flags interval
> > 	}
> > 
> > 	chain blacklist {
> > 		type filter hook input priority -10; policy accept;
> > 		ip saddr @attackers drop
> > 	}
> > }
> > ```
> 
> Actually I ment what happens when you delete the VyOS firewall - or update a rule. Will it influence sshguard?

As I see it not affected if you delete/change firewall rules
```
yos@tstrtr2# delete interfaces ethernet eth1 firewall 
[edit]
vyos@tstrtr2# compare 
-firewall {
-    name FOO {
-        rule 10 {
-            action accept
-        }
-    }
-}
[edit interfaces ethernet eth1]
-firewall {
-    local {
-        name FOO
-    }
-}
[edit]
vyos@tstrtr2# commit
[edit]
vyos@tstrtr2# sudo nft list table ip sshguard
table ip sshguard {
	set attackers {
		type ipv4_addr
		flags interval
	}

	chain blacklist {
		type filter hook input priority filter - 10; policy accept;
		ip saddr @attackers drop
	}
}
[edit]
vyos@tstrtr2# 
```

**Pull request merged**